third_party/move/move-bytecode-verifier/src/type_safety.rs

```diff
@@ -300,6 +301,115 @@ fn call(
     Ok(())
 }
 
+fn clos_eval(
+    verifier: &mut TypeSafetyChecker,
+    meter: &mut impl Meter,
+    offset: CodeOffset,
+    expected_ty: &SignatureToken,
+) -> PartialVMResult<()> {
+    let SignatureToken::Function(param_tys, ret_tys, abilities) = expected_ty else {
+        // The signature checker has ensured this is a function
+        safe_assert!(false);
+        unreachable!()
+    };
+    // On top of the stack is the closure, pop it.
+    let closure_ty = safe_unwrap!(verifier.stack.pop());
+    // Verify that the closure type matches the expected type
+    if &closure_ty != expected_ty {
```

Comment:
How do we implement `Eq` for signature token? Because later you do `!abilities.is_subset(closure_abilities)` but we check for equality here?

Reply:
Yea I would be a bit concerned about the usability of code around co-variance here. The code here will be a lot more painful to modify after we ship the feature so really want us to have a clear story here. Will the lack of covariance here actually hurt the usability?

Reply:
Folks, these are all standard patterns. Rust doesn't have co or contra variance, and multiple other languages too. Still all have substitutability/assignability. Please read the AIP also for discussion.

Rust e.g. considers `R<&T>` and `R<&mut T>` completely different types -- even though it allows to assign `&mut T` to `&T`.

Also not adding co-variance now will not block us from adding it later if we really wanted.

Reply:
I was more talking about the code here. If we already checked `closure_ty == expected_ty`, by using the default Eq implementation we are requiring the ability of the two closure types to be the same as well, which is not what you would want for assignability?

Reply:
Indeed! I have forgotten to update this file, its still in the state from October.

Now introduced `token.is_assignable_from(other)` and applied at the appropriate locations here. Also documented where should not be applied and equality used instead.

Also removed the unnecessary ability check in `call_closure`.